### DISCUSSION TITLE
How much of the expected feature set is completed?

### DISCUSSION THREAD

**Discussion by seivan** [dtzxporter/hydra#11] [General]
First of, amazing work on the API design, you clearly understand the assignment on what's missing for Rust. 

I believe this should be the foundation for most of the backend frameworks from now on, it would give the same ecosystem benefits that Elixir had for Phoenix. 

But I have a few questions... coming from Elixir, how many of the features are functional and stable now? 

1. Does gossiping for node connections work? 
2. Is `global` available for cluster wide registration of GenServers? 
3. Can `Server A` send a message to a `GenServer` created on `Server B` to mutate state, and expect a response? 
4. Can `Server A` monitor a `GenServer` on `Server B` to relaunch if it fails? 
5. Can I use [k8 headless services](https://github.com/bitwalker/libcluster/pull/36) or ECS DNS lookups to connect Node 6 to Node 5, and get Nodes 1-4 as well? 

Just trying to wrap my head around what features are missing and how stable the networking features are. 

In fact most of my concerns are the networking stuff, if the distributed processes have any issues.

**Comment by dtzxporter** [comment]
1. Yes, though no connection is made until an attempt to send to the discovered node (or monitor, etc) eventually will be configurable like erlang.
2. No, registries are available at the moment and you can use named processes to talk to them remotely.
3. Yes, works fully.
4. It can monitor it, but you would need some logic to restart it, you can't 'spawn' random code remotely of course because rust isn't dynamic.
5. You can manually manage nodes using the Node methods: https://docs.rs/hydra/latest/hydra/enum.Node.html#method.connect

**Reply by seivan** [reply]
> 1. Yes, though no connection is made until an attempt to send to the discovered node (or monitor, etc) eventually will be configurable like erlang.

Well, a connection should be made on `connect`, and the gossiping should make sure other nodes in the cluster are also visible. Having to pre-configure it kills a little bit of the dynamic element if you have auto-scaling. 

That's how it works in Elixir anyway, Node.connect `B` should also include nodes connected to the clusters. 
So `Node.list` would include `[B, C, D]` (when called on `A`). 

> 2. No, registries are available at the moment and you can use named processes to talk to them remotely.
> 3. Yes, works fully.

How does that actually look and work right now? 
With Elixir/OTP you can start a `GenServer` dynamically with `DynamicSupervisor`, have it be named `lobby-123` and be registered globally in the cluster using any of the `pg`, or `:global` registries. This is also safe from race-conditions where GenServer can `find_or_create`. 

So starting something like `{:via, :global, lobby-123}` would either create it, and if it exists would just return thats pid. Obviously the `init` would have to manage that, just like in Elixir. 

Any `cast/call` to `{:via, :global, lobby-123}` will be redirected to that `GenServer`  - useful for multiple nodes running behind a load balancer, not to have to worry about sticky-sessions. 

How would that look like with `Hydra`, and what are the missing steps?

**Comment by dtzxporter** [comment]
> Well, a connection should be made on connect, and the gossiping should make sure other nodes in the cluster are also visible. Having to pre-configure it kills a little bit of the dynamic element if you have auto-scaling.

This is subjective of course, many people provide their own service discovery system built on elixir, and explicitly turn off auto connect! Regardless, this will be a feature just like it is in elixir/erlang. It's easy to implement, I just need to test a few cases first before shipping it, since I don't run this in production.

Registries exist, but you'll need to explicitly call them using their named registry:
```
// Call a remote registry with name, @ the node.
Registry.lookup_or_start((name, node), key)...
// Call a known local registry:
Registry.lookup_or_start(name, key)...
```

(This registry is separate to the 'spawn via' that's available in elixir, but an api will be available once a global registry exists)

I do plan on creating a global registry that is sync'd between connected nodes at some point, but it's currently not implemented.

Syntax will probably be like:
`cast((Via, "registry name", "key"), ...)`

More to come shortly! I'm actively involved in several projects taking advantage of hydra, and a lot of features will be polished up in the new year.
